doc/sphinx/proof-engine/tactics.rst

```diff
@@ -650,180 +649,339 @@ Applying theorems
+
+      .. coqtop:: all
 
-   .. tacv:: eapply @term
-      :name: eapply
+         apply H1 in H0.  (* change H0, create new goals for unmatched premises of H1 *)
 
-      The tactic :tacn:`eapply` behaves like :tacn:`apply` but it does not fail when no
-      instantiations are deducible for some variables in the premises. Rather, it
-      turns these variables into existential variables which are variables still to
-      instantiate (see :ref:`Existential-Variables`). The instantiation is
-      intended to be found later in the proof.
+   .. _apply_with_binding_goal:
+   .. example:: Apply a theorem with a binding in a goal
 
-   .. tacv:: rapply @term
-      :name: rapply
+      :tacn:`apply` unifies the conclusion `n <= p` of the theorem
+      `le_trans : forall n m p, n <= m -> m <= p -> n <= p`
+      with the goal, assigning `x * x ^ n` and `y * y ^ n` in the goal
+      to, repectively, `n` and `p` in theorem (backward reasoning).
+      The `with` clause provides the binding for `m`:
 
-      The tactic :tacn:`rapply` behaves like :tacn:`eapply` but it
-      uses the proof engine of :tacn:`refine` for dealing with
-      existential variables, holes, and conversion problems.  This may
+      .. coqtop:: reset none
+
+         Require Import PeanoNat.
+         Require Import Nat.
+         Infix "^" := pow : nat_scope.
+         Goal forall (x y n : nat), x * x ^ n <= y * y ^ n.
+
+      .. coqtop:: out
+
+         intros x y n.
+
+      .. coqtop:: all
+
+         apply Nat.le_trans with (y * x ^ n).
+
+   .. _apply_with_binding_hyp:
+   .. example:: Apply a theorem with a binding in a hypothesis
+
+      When applying a theorem in a hypothesis,
+      :tacn:`apply` unifies the hypothesis with one of the premises
+      of the theorem.  In this case, it unifies with the first premise
+      (`n <= m`) and assigns `x * x ^ n` and `y * y ^ n` to,
+      respectively, `n`` and `m` in the theorem (forward reasoning).
+      The  `with` clause provides the binding for `p`.
+
+      In addition,
+      `apply` in a hypothesis isn't as flexible as
+      `apply` in the goal: for hypotheses, the unbound variable can be bound
+      by name (as shown) or values for all the variables can be given
+      positionally,
+      i.e. `apply Nat.le_trans with (x * x ^ n) (y * y ^ n) (y * x ^ n) in H.`
+
+      .. coqtop:: reset none
+
+         Require Import PeanoNat.
+         Require Import Nat.
+         Infix "^" := pow : nat_scope.
+         Goal forall (x y n : nat) (H: x * x ^ n <= y * y ^ n), True.
+
+      .. coqtop:: out
+
+         intros x y n H.
+
+      .. coqtop:: all
+
+         apply Nat.le_trans with (p := y * x ^ n) in H.
+
+   .. _apply_with_iff:
+   .. example:: Applying theorems with `<->`
+
+      Theorems that use `<->` to state a logical equivalence behave
+      consistently when applied to goals and hypotheses; the tactic
+      picks the useful alternative.  The lemma
+      is `iff: 1 = 1 <-> 2 = 2`:
```

Comment:
What about adding this "lemma" to the context instead of having to present it with text?